core/string/translation.cpp

```diff
@@ -1300,6 +1316,309 @@ StringName TranslationServer::doc_translate_plural(const StringName &p_message,
+}
+
+void TranslationServer::set_pseudolocalization_expansion_ratio(float ratio) {
+	expansion_ratio = ratio;
+
+	if (OS::get_singleton()->get_main_loop()) {
+		OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_TRANSLATION_CHANGED);
+	}
+	ResourceLoader::reload_translation_remaps();
+}
+
+String TranslationServer::get_pseudolocalization_prefix() const {
+	return pseudolocalization_prefix;
+}
+
+void TranslationServer::set_pseudolocalization_prefix(String prefix) {
+	pseudolocalization_prefix = prefix;
+
+	if (OS::get_singleton()->get_main_loop()) {
+		OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_TRANSLATION_CHANGED);
+	}
+	ResourceLoader::reload_translation_remaps();
+}
+
+String TranslationServer::get_pseudolocalization_suffix() const {
+	return pseudolocalization_suffix;
+}
+
+void TranslationServer::set_pseudolocalization_suffix(String suffix) {
+	pseudolocalization_suffix = suffix;
+
+	if (OS::get_singleton()->get_main_loop()) {
+		OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_TRANSLATION_CHANGED);
+	}
+	ResourceLoader::reload_translation_remaps();
+}
+
+StringName TranslationServer::pseudolocalize(const StringName &p_message) const {
+	String message = p_message;
+	int length = message.length();
+	if (pseudolocalization_override_enabled) {
+		message = get_override_string(length);
+	}
+
+	if (pseudolocalization_double_vowels_enabled) {
+		message = double_vowels(message);
+	}
+
+	if (pseudolocalization_accents_enabled) {
+		message = replace_with_accented_string(message);
+	}
+
+	if (pseudolocalization_fake_bidi_enabled) {
+		message = wrap_with_fakebidi_characters(message);
+	}
+
+	StringName res = add_padding(message, length);
+	return res;
+}
+String TranslationServer::get_override_string(int length) const {
+	String res = "";
+	for (int i = 0; i < length; i++) {
+		res += '*';
+	}
+	return res;
+}
+String TranslationServer::double_vowels(String &message) const {
+	String res = "";
+	for (int i = 0; i < message.size(); i++) {
+		res += message[i];
+		if (message[i] == 'a' || message[i] == 'e' || message[i] == 'i' || message[i] == 'o' || message[i] == 'u' ||
+				message[i] == 'A' || message[i] == 'E' || message[i] == 'I' || message[i] == 'O' || message[i] == 'U') {
+			res += message[i];
+		}
+	}
+	return res;
+};
+
+String TranslationServer::replace_with_accented_string(String &message) const {
+	String res = "";
+	for (int i = 0; i < message.size(); i++) {
+		const wchar_t *accented = get_accented_version(message[i]);
+		if (accented) {
+			res += accented;
+		} else {
+			res += message[i];
+		}
+	}
+	return res;
+}
+
+String TranslationServer::wrap_with_fakebidi_characters(String &message) const {
+	String res = "";
+	wchar_t fakebidiprefix = L'\u202e';
+	wchar_t fakebidisuffix = L'\u202c';
+	res += fakebidiprefix;
+	//the fake bidi unicode gets popped at every newline so pushing it back at every newline.
+	for (int i = 0; i < message.size(); i++) {
```

Comment:
It might be useful to pop override around formatting string parts (`%s` etc.), e.g. make it `(RLO)text text (PDF)%s(RLO) text text(PDF)` instead of `(RLO)text text %s text text(PDF)`.

This will help to identify strings which need additional control chars to display correctly, e.g. list of node names (not translated) in the middle of TRL text, like this one https://hosted.weblate.org/translate/godot-engine/godot/he/?checksum=3818bad5b65a4bef

Reply:
So, should I just pop the FakeBiDi for formatting strings part or skip pseudolocalizing them altogether? Also, I should do this for all %s, %d, etc or also for {s} kind of parts?

Reply:
Just pop it for `%s` (and other string type formatting if there are any).